GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,274 @@
+            }
+
+            return path.Substring(0, whackLocation);
+        }
+
+        /// <summary>
+        /// Categorizes a list of paths given as strings by mapping them to the top level directory in their path
+        /// Modifies the directoryTracking dictionary to have an accurate count of the files underneath a top level directory
+        /// </summary>
+        /// <remarks>
+        /// The distinction between files and directories is important --
+        /// If the path to a file doesn't contain a GVFSConstants.GitPathSeparator, then that means it falls within the root
+        /// However if a directory's path doesn't contain a GVFSConstants.GitPathSeparator, it doesn't count towards its own hydration
+        /// </remarks>
+        /// <param name="paths">An enumerable containing paths as strings</param>
+        /// <param name="directoryTracking">A dictionary used to track the number of files per top level directory</param>
+        /// <param name="parentDirectory">Paths will only be categorized if they are descendants of the parentDirectory</param>
+        /// <param name="isFile">Flag which specifies whether the list of paths are paths to files (true) or directories (false)</param>
+        private int CategorizePaths(IEnumerable<string> paths, Dictionary<string, int> directoryTracking, string parentDirectory, bool isFile)
+        {
+            int count = 0;
+            foreach (string path in paths)
+            {
+                // Only categorize if descendent of the parentDirectory
+                if (path.StartsWith(parentDirectory))
+                {
+                    count++;
+
+                    if (isFile)
+                    {
+                        // Find out the top level directory of the files path, which will be one level under the parent directory
+                        // If the file is in the parent directory, topDir is set to '/' (just temporary, formatting will eventually just show the path to the parent directory)
+                        string topDir = this.ParseTopDirectory(this.TrimDirectoryFromPath(path, parentDirectory));
+                        this.IncreaseDictionaryCounterByKey(directoryTracking, topDir);
+                    }
+                    else
+                    {
+                        // If the path is to the parentDirectory, ignore it to avoid adding string.Empty to the data structures
+                        if (!parentDirectory.Equals(path))
+                        {
+                            // Trim the path to parent directory from the path to this directory
+                            string topDir = this.TrimDirectoryFromPath(path, parentDirectory);
+
+                            // If the directory isn't one level underneath the parent, count it
+                            // Otherwise, don't add it to the dictionary (it shouldn't count itself)
+                            if (topDir.IndexOf(GVFSConstants.GitPathSeparator) != -1)

```

Comment:
What type of path would result in a `false` value for this condition?

Reply:
The path to a directory one level underneath the parent directory